### ISSUE TITLE
Customization of All DataStax Settings

### ISSUE THREAD

**Issue by pluttrell** [Impetus/kundera#724]
I am migrating from using Cassandra and DataStax directly and have heavily customized the configuration of our Cluster instance. I see that Kundera has a mechanism to supply the underlying DataStax drivers with custom configuration [https://github.com/impetus-opensource/Kundera/wiki/Datastax-java-driver-support-for-Cassandra-using-Kundera](https://github.com/impetus-opensource/Kundera/wiki/Datastax-java-driver-support-for-Cassandra-using-Kundera), but doesn't support all of the options that DataStax does. 

Here is an example of some of our code, note that we do things like wrap the round robin policies for load balancing:

```
PoolingOptions poolingOptions = new PoolingOptions();
poolingOptions.setMaxConnectionsPerHost(HostDistance.LOCAL, maxConnectionsPerLocalHost);
poolingOptions.setCoreConnectionsPerHost(HostDistance.LOCAL, coreConnectionsPerLocalHost);
poolingOptions.setMaxSimultaneousRequestsPerConnectionThreshold(HostDistance.LOCAL, maxSimultaneousRequestsPerLocalConnectionThreshold);
poolingOptions.setMaxSimultaneousRequestsPerHostThreshold(HostDistance.LOCAL, maxSimultaneousRequestsPerLocalHostThreshold);
poolingOptions.setMinSimultaneousRequestsPerConnectionThreshold(HostDistance.LOCAL, minSimultaneousRequestsPerLocalConnectionThreshold);
poolingOptions.setPoolTimeoutMillis(poolTimeoutMillis);

String[] addresses = ...

Cluster cluster = Cluster.builder()
    .addContactPoints(addresses)
    .withPort(serverConnectionPort)
    .withCredentials(cassandraUsername, cassandraPassword)
    .withCompression(compressionMethod)
    .withLoadBalancingPolicy(new TokenAwarePolicy(new DCAwareRoundRobinPolicy()))
    .withReconnectionPolicy(new ExponentialReconnectionPolicy(maxReconnectionTimeInMillis))
    .withRetryPolicy(new FixedRetryPolicy(maxAttempts))
    .withSocketOptions(new SocketOptions()
        .setConnectTimeoutMillis(connectTimeoutInMillis)
        .setReadTimeoutMillis(readTimeoutInMillis))
    .withPoolingOptions(poolingOptions)
    .build();
```

Is there a way I can supply Kundera with the `Cluster` and/or `Session` instance we already use? If not, how can we customize _all_ options in the DataStax driver?

**Comment by karthikprasad13** [comment]
@pluttrell 
Yes, Kundera supports customization of cluster by defining the settings in an external property file.
Please do the following to achieve the same:

1) Set the properties in an external xml file. You can refer [these](https://github.com/impetus-opensource/Kundera/tree/c6101f21cd8f079fb4cc44e85de49261bf5fef7e/src/kundera-cassandra/cassandra-ds-driver/src/test/resources) property files as examples.
    Also, I hope you referred this [test case](https://github.com/impetus-opensource/Kundera/blob/trunk/src/kundera-cassandra/cassandra-ds-driver/src/test/java/com/impetus/kundera/client/cassandra/dsdriver/DSClientExternalPropertyTest.java), on how to set properties using property map.

2) Add the following property in your `persistence.xml`
        `<property name="kundera.client.property" value="dsclient.xml" />`

HTH
Karthik

**Comment by pluttrell** [comment]
Thank you for providing links to the current samples and unit tests for the Kundera configuration strategy.

What I was looking for however was if there was an ability to not utilize the Kundera configuration strategy and instead provide the `Cluster` or `Session` instances directly to Kundera?

Here's why I ask: We've configured things at the DataStax driver level that the Kundera code doesn't appear to do. For example notice in my above example the following:
- The LoadBalancingPolicy we use is the `DCAwareRoundRobinPolicy` wrapped in a `TokenAwarePolicy`. Such a configuration is documented by DataStax, but doesn't appear to be supported by Kundera as it uses reflection to load a class, which is defined by it's property file. You'd need to provide parameters to the constructor, which of course makes the property file much more complex.
- The ReconnectionPolicy, where we use the `ExponentialReconnectionPolicy`, but we send in the maxReconnectionTime. Kundera's configuration strategy doesn't appear to provide support for this.
- The RetryPolicy, where we use our own custom implementation, which doesn't appear to be supported by Kundera's configuration strategy.

DataStax will very likely continue to evolve their configuration options as we've seen with the V2/V3 specific options... and with the Kundera's current strategy, Kundera would likely need to track these changes release-by-release. Long term, it might be a good idea to support a means to inject the `Cluster` or `Session` instance to handle cases where the Kundera configuration strategy doesn't meet a companies needs.

**Comment by karthikprasad13** [comment]
@pluttrell 

Kundera does not support injecting `Cluster` or `Session` instances directly as of now because of the complications that need to be addressed at the transactional level. Your point is very valid that Kundera needs to track the changes release-by-release because of this. We are looking into it.

Regarding the configurations, most of them are supported by Kundera. You can specify them in the external property file like I said in my previous comment.
- "LoadBalancingPolicy we use is the `DCAwareRoundRobinPolicy` wrapped in a `TokenAwarePolicy`" => You can wrap any child policy in a `TokenAwarePolicy` by adding the following property in your property file:
  
  ```
              <property name="isTokenAware" value="true"></property>
  ```
- "we use the ExponentialReconnectionPolicy, but we send in the maxReconnectionTime" => This is possible with Kundera. You can specify the maxReconnectionTime (say 13000 ms)  using the following property:
  
  ```
             <property name="maxDelayMs" value="13000"></property>
  ```
- Using custom `RetryPolicy` is not currently supported in Kundera and will be planned soon.

Also, We are not using any reflections in our implementation. Hence, in order to set policies using Kundera's configuration strategy, you just need to provide the property in the xml file and Kundera will handle the configurations internally.

Karthik.

**Comment by pluttrell** [comment]
@karthikprasad13 

Thanks for getting back to me and explaining how Kundera can support a couple of the examples I asked about. It would be great if these (and any other missing properties) could be added to the documentation.

Regarding not having a way to specify the `RetryPolicy`, it would be great if this could be added. Lack of support for this was one of the reasons that the team I am advising can't adopt Kundera at this time.

Thanks for clarification on the use of reflection. I miss-assumed that Kundera used reflection to load configuration option classes because of the full class names in `String` fields in the "External Properties Configuration" section on [this](https://github.com/impetus-opensource/Kundera/wiki/Datastax-java-driver-support-for-Cassandra-using-Kundera) page in the docs.

It might be a good idea to re-consider providing a means to inject a `Cluster` or `Session` instance and finding a solution to the transactionality complications. In our internal services framework, we essentially inject a `Cluster` instance and have no challenges regarding transactionality...thought we don't implement JPA, which is where perhaps the challenges for Kundera show up.

**Comment by karthikprasad13** [comment]
@pluttrell 

Thank you for your valuable suggestions. We have started working on the issues and will fix as soon as possible. Please let us know other problems (if any) that Kundera failed to solve in your usecase.

Karthik.

**Comment by karthikprasad13** [comment]
@pluttrell 

We have added the support for specifying custom `RetryPolicy`. Please refer [this](https://github.com/impetus-opensource/Kundera/blob/821a4a17a726e48d4f549e2193a6e265f0a2ca34/src/kundera-cassandra/cassandra-ds-driver/src/test/resources/dsclient.xml) property file. It shows how we can provide our custom retry policy to Kundera.

Karthik.

**Comment by mounowicz** [comment]
@karthikprasad13 

Hello, we have a similar problem, however with SSL support. We need to connect our distant Cassandra servers using SSL which works for us using the native driver, however I didn't find how it could be achieved using Kundera. Is it possible?

Many thanks.

Tomas
